controller/update.go

```diff
@@ -0,0 +1,126 @@
+package controller
+
+import (
+	commonauth "github.com/fabric8-services/fabric8-common/auth"
+	"github.com/fabric8-services/fabric8-common/errors"
+	"github.com/fabric8-services/fabric8-common/log"
+	"github.com/fabric8-services/fabric8-tenant/app"
+	"github.com/fabric8-services/fabric8-tenant/cluster"
+	"github.com/fabric8-services/fabric8-tenant/configuration"
+	"github.com/fabric8-services/fabric8-tenant/environment"
+	"github.com/fabric8-services/fabric8-tenant/jsonapi"
+	"github.com/fabric8-services/fabric8-tenant/openshift"
+	"github.com/fabric8-services/fabric8-tenant/update"
+	"github.com/fabric8-services/fabric8-wit/ptr"
+	"github.com/goadesign/goa"
+	"github.com/jinzhu/gorm"
+)
+
+var fabric8TenantUpdateSA = "fabric8-tenant-update"
+
+// UpdateController implements the update resource.
+type UpdateController struct {
+	*goa.Controller
+	db             *gorm.DB
+	config         *configuration.Data
+	clusterService cluster.Service
+	updateExecutor openshift.UpdateExecutor
+}
+
+// NewUpdateController creates a update controller.
+func NewUpdateController(service *goa.Service, db *gorm.DB, config *configuration.Data, clusterService cluster.Service, updateExecutor openshift.UpdateExecutor) *UpdateController {
+	return &UpdateController{
+		Controller:     service.NewController("UpdateController"),
+		db:             db,
+		config:         config,
+		clusterService: clusterService,
+		updateExecutor: updateExecutor}
+}
+
+// Start runs the start action.
+func (c *UpdateController) Start(ctx *app.StartUpdateContext) error {
+	if !commonauth.IsSpecificServiceAccount(ctx, fabric8TenantUpdateSA) {
+		return jsonapi.JSONErrorResponse(ctx, errors.NewUnauthorizedError("Wrong token"))
+	}
+
+	var envTypesFilter update.FilterEnvType
+	if value(ctx.EnvType) != "" {
+		envType, ok := interface{}(value(ctx.EnvType)).(environment.Type)
+		if !ok {
+			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("env-type", ctx.EnvType))
+		}
+		envTypesFilter = update.OneType(envType)
+	} else {
+		envTypesFilter = update.AllTypes
+	}
+
+	go update.NewTenantsUpdater(c.db, c.config, c.clusterService, c.updateExecutor, envTypesFilter, value(ctx.ClusterURL)).UpdateAllTenants()
+
+	return ctx.Accepted()
+}
+
+func value(ptr *string) string {
+	if ptr == nil {
+		return ""
+	}
+	return *ptr
+}
+
+// Show runs the show action.
+func (c *UpdateController) Show(ctx *app.ShowUpdateContext) error {
+	if !commonauth.IsSpecificServiceAccount(ctx, fabric8TenantUpdateSA) {
+		return jsonapi.JSONErrorResponse(ctx, errors.NewUnauthorizedError("Wrong token"))
+	}
+
+	var tenantsUpdate *update.TenantsUpdate
+	err := update.Transaction(c.db, func(tx *gorm.DB) error {
+		var err error
+		tenantsUpdate, err = update.NewRepository(tx).GetTenantsUpdate()
```

Comment:
We don't have to execute read only operations in transactions.

Reply:
In this particular case not, but in the rest of the cases, I need to read it in a transaction with the locked table. Not sure if there is some easy way to call the method without using a transaction when I need it in the other cases

Reply:
wait, I can actually do this: b7cf997